connector/src/main/scala/com/microsoft/kusto/spark/authentication/DeviceAuthentication.scala

```diff
@@ -1,115 +1,63 @@
-import org.apache.log4j.{Level, Logger}
-
-import scala.util.Try
-
-object DeviceAuthentication {
-  // This is the Kusto client id from the java client used for device authentication.
-  private val CLIENT_ID = "db662dc1-0cfe-4e1c-a843-19a68e65be58"
-
-  def acquireDeviceCode(clusterUrl: String, authority: String = "common"): DeviceCode = {
-    var aadAuthorityUri = ""
-    val aadAuthorityFromEnv = System.getenv("AadAuthorityUri")
-    if (aadAuthorityFromEnv == null) {
-      aadAuthorityUri = String.format("https://login.microsoftonline.com/%s", authority)
-    } else {
-      aadAuthorityUri = String.format("%s%s%s", aadAuthorityFromEnv, if (aadAuthorityFromEnv.endsWith("/")) "" else "/", authority)
-    }
-
-    val service = Executors.newSingleThreadExecutor
-    val context: AuthenticationContext =  new AuthenticationContext(aadAuthorityUri, true, service)
-
-    context.acquireDeviceCode(CLIENT_ID, clusterUrl, null).get
+import java.util.concurrent.CompletableFuture
+import java.util.function.Consumer
+
+import com.microsoft.aad.msal4j.{DeviceCode, DeviceCodeFlowParameters, IAuthenticationResult}
+import com.microsoft.azure.kusto.data.auth
+
+import scala.concurrent.TimeoutException
+
+class DeviceAuthentication (val cluster: String, val authority:String) extends auth.DeviceAuthTokenProvider(cluster, authority) {
+  var deviceCode: Option[DeviceCode] = None
+  var expiresAt: Option[Long] = None
+  var awaitAuthentication: Option[CompletableFuture[IAuthenticationResult]] = None
+  val NEW_DEVICE_CODE_FETCH_TIMEOUT = 5000
+  override def acquireNewAccessToken(): IAuthenticationResult = {
+    awaitAuthentication = Some(acquireNewAccessTokenAsync())
+    awaitAuthentication.get.join()
   }
 
-  def acquireAccessTokenUsingDeviceCodeFlow(clusterUrl: String, authority: String = "common", userDeviceCode: Option[DeviceCode] = None) : String = {
-    val deviceCode = if (userDeviceCode.isDefined) userDeviceCode.get else acquireDeviceCode(clusterUrl, authority)
-    val aadAuthorityUri = s"https://login.microsoftonline.com/$authority"
-    val service: ExecutorService =
-      Executors.newSingleThreadExecutor
-    val context: AuthenticationContext =  new AuthenticationContext(aadAuthorityUri, true, service)
-
-    var text: String = null
-    Try {
-      val clipboard = Toolkit.getDefaultToolkit.getSystemClipboard
-      val dataFlavor = DataFlavor.stringFlavor
-      if (clipboard.isDataFlavorAvailable(dataFlavor)) {
-        text = clipboard.getData(dataFlavor).asInstanceOf[String]
-      }
-      clipboard.setContents(new StringSelection(deviceCode.getUserCode), null)
+  def acquireNewAccessTokenAsync(): CompletableFuture[IAuthenticationResult] = {
+    val deviceCodeConsumer: Consumer[DeviceCode] = (deviceCode: DeviceCode) => {
+      this.deviceCode = Some(deviceCode)
+      expiresAt = Some(System.currentTimeMillis + deviceCode.expiresIn() * 1000)
+      System.out.println(deviceCode.message())
```

Comment:
Why not just "println"? AFAIU, they're different in Scala

Reply:
just cause I copied from java code .. ill change